### ISSUE TITLE
Contact Birhtday events change etag all the time despite being identical

### ISSUE THREAD

**Issue by Garfonso** [owncloud/contacts#997]
I did not see this issue in owncloud 8.0.4 and am pretty sure it is new in owncloud 8.1.
Not sure if this is more a contacts or a calendar issue... but I know that the calendar events for contact birthdays are generated in the contacts app, so I thought I'd try my luck over here.

### Steps to reproduce
1. fill birthday fields in some contacts
2. Send the following caldav request:
REPORT
Prefer: return-minimal
Content-Type: text/xml; charset=utf-8
Authorization: Basic auth
Depth: 1
Content-Length: 316
<?xml version="1.0" encoding="utf-8" ?><c:calendar-query xmlns:d="DAV:" xmlns:c="urn:ietf:params:xml:ns:caldav"><d:prop><d:getetag /></d:prop><c:filter><c:comp-filter name="VCALENDAR"><c:comp-filter name="VEVENT"><c:time-range start="20140720T000000Z" /></c:comp-filter></c:comp-filter></c:filter></c:calendar-query>
to the server https:/***/remote.php/caldav/calendars/USERNAME/contact_birthdays/
3. See that the etag fields in the result xml change for every single request.

### Expected behaviour
The etag field should only change if the event changes. This was correct before and still works for all other calendars. Also the ctag field works fine for the whole contact birthdays calendar, but etags for the single objects do not work at all. :-(

### Actual behaviour
The etag field for single events (i.e. "ics files") changes every time a new report / calendar query is sent. This makes clients think that all birthday events did change and they need to be downloaded again. But, obviously, they did not change at all. So this is just a waste of energy and bandwidth (which hurts especially on mobile clients).

### Server configuration
**Operating system**:  Ubuntu 14.04, 64Bit
Linux Server 3.13.0-58-generic #97-Ubuntu SMP Wed Jul 8 02:56:15 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

**Web server:**
Apache/2.4.7 (Ubuntu) PHP/5.5.9-1ubuntu4.11 OpenSSL/1.0.1f

**Database:**
mysql  Ver 14.14 Distrib 5.5.43, for debian-linux-gnu (x86_64) using readline 6.3

**PHP version:**
PHP 5.5.9-1ubuntu4.11 (cli) (built: Jul  2 2015 15:23:08) 
Copyright (c) 1997-2014 The PHP Group
Zend Engine v2.5.0, Copyright (c) 1998-2014 Zend Technologies
    with Zend OPcache v7.0.3, Copyright (c) 1999-2014, by Zend Technologies

**ownCloud version:** (see ownCloud admin page)
ownCloud 8.1.0 (stable) 

**Updated from an older ownCloud or fresh install:**
updated (since at least 6.0)

**List of activated apps:**
Enabled:
 - activity: 2.0.1
 - appframework: 0.104
 - calendar: 0.7.2
 - contacts: 0.4.0.0
 - files: 1.1.9
 - files_antivirus: 0.7.0.1
 - files_external: 0.2.3
 - files_locking: 
 - files_pdfviewer: 0.7
 - files_sharing: 0.6.2
 - files_texteditor: 0.4
 - files_trashbin: 0.6.3
 - files_versions: 1.0.6
 - files_videoviewer: 0.1.3
 - firstrunwizard: 1.1
 - gallery: 0.6.0
 - mail: 0.1.3
 - search_lucene: 0.6.0
 - user_external: 0.4
Disabled:
 - encryption
 - external
 - mozilla_sync
 - news
 - provisioning_api
 - roundcube
 - templateeditor
 - user_ldap
 - user_webdavauth

**The content of config/config.php:**
sudo -u www-data php occ config:list system --public => does not work for me, result:   [InvalidArgumentException] There are no commands defined in the "config" namespace. 

Manually cleaned paste:
<?php
$CONFIG = array (
'instanceid' => '',
'passwordsalt' => '',
'datadirectory' => '/var/owncloud/data',
'dbtype' => 'mysql',
'dbname' => '**',
'dbuser' => '',
'dbpassword' => '',
'dbhost' => '',
'dbtableprefix' => 'oc_',
'version' => '8.1.0.8',
'installed' => true,
'forcessl' => true,
'theme' => '',
'maintenance' => false,
'trusted_domains' =>
array (
0 => '',
1 => '',
2 => '',
),
'loglevel' => '0',
'mail_from_address' => 'owncloud',
'mail_smtpmode' => 'php',
'mail_domain' => 'tratschtante.de',
'secret' => '****',
'overwrite.cli.url' => '/owncloud',
'preview_libreoffice_path' => '/usr/bin/libreoffice',
'cron_log' => true,
'appstoreenabled' => true,
'has_internet_connection' => true,
'apps_paths' =>
array (
0 =>
array (
'path' => '/var/www/owncloud/apps',
'url' => '/apps',
'writable' => true,
),
),
'memcache.local' => '\OC\Memcache\APCu',
);

**Are you using external storage, if yes which one:** local/smb/sftp/...
local
smb

**Are you using encryption:** yes/no
no

**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...
no

### Client configuration
**Browser:**
node.js script

**Operating system:**
webOS, Windows 8.1, Ubuntu 14.04 64Bit all behave the same

#### Web server error log
empty (i.e. no new entries show up during the request)

#### ownCloud log (data/owncloud.log)
empty (i.e. no new entries show up during the request)

#### Browser log
Uhm... I could paste you a few versions of the xml? But I'm pretty sure that'd be of not much help.

**Comment by gvde** [comment]
When the birthday event is generated it does not set DTSTAMP. By default, an event uses the current time as DTSTAMP time thus the calendar event has a different DTSTAMP each time and thus the etag generated from the event content changes with each time.
Please try commit e172aa8c08dead75338ff2c10ecff5807367e40e and see if that works for you. It uses the last modification time of the contact as DTSTAMP timestamp, i.e. the event will only change if the contact has been changed.

**Comment by Garfonso** [comment]
Sorry this took me some time to test.The commit solves the issue with the etags. Nice. :+1: Thank you for that. 

But this commit does some strange things with TZID, it produces lines like this in the ics:
DTSTAMP;TZID="+00:00":20150220T111319
Having a ":" in a quoted string is allowed. But this is not a valid TZID (it probably would be if there would be a VTIMEZONE section for "+00:00" in the VCALENDAR). 
One way to go would be to get rid of the timezone and use UTC time, which is commonly done for the DTSTAMP. Or it is necessary to tell this the correct local TZ and set the right TZID. My gut tells me this should be a parameter to the setDateTime function you call in the code. But I haven't looked it up in the docs or something.

**Comment by gvde** [comment]
I didn't bother for tests to set the time zone. My first quick attempt didn't work and I didn't have the time to find out...
See 4c18e053faff748d3864a3ab7f80d7d992c8ba3b to set the timezone to UTC a

**Comment by gvde** [comment]
Created pull requests #1005 and #1006

**Comment by Garfonso** [comment]
Yeah, UTC looks a lot better. Sorry, did not manage to test yesterday. 

Great. Thanks. :+1:
